cmd/node_builder.go

```diff
@@ -0,0 +1,128 @@
+)
+
+// NodeBuilder declares the initialization methods needed to bootstrap up a Flow node
+type NodeBuilder interface {
+
+	// BaseFlags reads the command line arguments common to all nodes
+	BaseFlags()
+
+	// ExtraFlags reads the node specific command line arguments and adds it to the FlagSet
+	ExtraFlags(f func(*pflag.FlagSet)) NodeBuilder
+
+	// ParseAndPrintFlags parses all the command line arguments
+	ParseAndPrintFlags()
+
+	// Initialize performs all the initialization needed at the very start of a node
+	Initialize() NodeBuilder
+
+	// PrintBuildVersionDetails prints the node software build version
+	PrintBuildVersionDetails()
+
+	// EnqueueNetworkInit enqueues the default network component
+	EnqueueNetworkInit()
+
+	// EnqueueMetricsServerInit enqueues the metrics component
+	EnqueueMetricsServerInit()
+
+	// Enqueues the Tracer component
+	EnqueueTracer()
+
+	// Module enables setting up dependencies of the engine with the builder context
+	Module(name string, f func(builder NodeBuilder) error) NodeBuilder
+
+	// Component adds a new component to the node that conforms to the ReadyDone
+	// interface.
+	//
+	// When the node is run, this component will be started with `Ready`. When the
+	// node is stopped, we will wait for the component to exit gracefully with
+	// `Done`.
+	Component(name string, f func(NodeBuilder) (module.ReadyDoneAware, error)) NodeBuilder
+
+	// MustNot asserts that the given error must not occur.
+	// If the error is nil, returns a nil log event (which acts as a no-op).
+	// If the error is not nil, returns a fatal log event containing the error.
+	MustNot(err error) *zerolog.Event
+
+	// Run initiates all common components (logger, database, protocol state etc.)
+	// then starts each component. It also sets up a channel to gracefully shut
+	// down each component if a SIGINT is received.
+	Run()
+
+	// PreInit registers a new PreInit function.
+	// PreInit functions run before the protocol state is initialized or any other modules or components are initialized
+	PreInit(f func(node NodeBuilder)) NodeBuilder
+
+	// PostInit registers a new PreInit function.
+	// PostInit functions run after the protocol state has been initialized but before any other modules or components
+	// are initialized
+	PostInit(f func(node NodeBuilder)) NodeBuilder
+
+	// RegisterBadgerMetrics registers all badger related metrics
+	RegisterBadgerMetrics()
+
+	// getters
+	Config() BaseConfig
+	NodeID() flow.Identifier
+	Logger() zerolog.Logger
+	Me() *local.Local
+	Tracer() module.Tracer
+	MetricsRegisterer() prometheus.Registerer
+	Metrics() Metrics
+	DB() *badger.DB
+	Storage() Storage
+	ProtocolEvents() *events.Distributor
+	ProtocolState() protocol.State
+	Middleware() *p2p.Middleware
+	Network() *p2p.Network
+	MsgValidators() []network.MessageValidator
+	FvmOptions() []fvm.Option
+	NetworkKey() crypto.PrivateKey
+	RootBlock() *flow.Block
+	RootQC() *flow.QuorumCertificate
+	RootSeal() *flow.Seal
+	RootChainID() flow.ChainID
+
```

Comment:
> I wonder if it would be possible to put these into a separate data struct, rather than together with the interface, and have all implementations embed both?

Another possibility is to have a single method that returns some `NodeData` struct, which contains all of these fields. This would still not be entirely ideal, but might be an easier change to make for now?

I think eventually, it might make sense to separate the `NodeBuilder` interface (the functionality) from the data entirely, having a separate struct encapsulate the latter. The methods `Module`, `Component`, `PostInit`, etc. could then be passed in `f func(builder NodeBuilder, data NodeData) error` instead of `f func(builder NodeBuilder) error`.

This can probably be in a separate PR though.

Reply:
While finessing the getter/setter approach may be a bit premature for now, I think:
- it's [ok](https://golang.org/doc/effective_go#Getters) if not quite idiomatic,
- we're defining the interface at a time where we do have numerous instances for it, which is the one thing Go does recommend,

so I'm tempted to defer to @vishalchangrani 's good sense. I would point out though that this is already a big PR, and probably daunting to other reviewers => any change that reduces the # of SLoC, even at some marginal cost in flexibility of this interface, hastens the moment of merge.

Reply:
Thanks for that excellent suggestion @smnzhu. I did make that change and now have a `NodeConfig` struct which holds the data that is modified (removing all the Get/Set from the `NodeBuilder` interface).
The change actually reduces the number of lines changed since the existing code uses `node.dot.xxx` everywhere and now is now just `NodeConfig`